### ISSUE TITLE
mlflow run should ensure that mlflow is available inside the resultant conda environment

### ISSUE THREAD

**Issue by aarondav** [mlflow/mlflow#100]
Any project which I run using `mlflow run` should presumably have `mlflow` installed. Currently, we might recommend that people establish a pip dependency in their conda environment to include mlflow. This comes with a few pitfalls:

  - There is no connection between the `mlflow run` version and the version of `mlflow` in the conda environment. This discrepancy might cause certain API incompatibility (for example if the wrapper script starts or does not start a run, as discussed in #82). Additionally, I may update mlflow but forget to update the conda environment files of all projects.
  - Our mlflow caching mechanism does not work if I have a pip dependency without a version. If the version is updated on pypi, our cache is invalid, but I would have to manually delete the `mlflow-$sha` conda environment in order to invalidate it.

We want users to be able to easily upgrade mlflow and not run into compatibility problems, so I would recommend that we automatically inject the mlflow version of the outer script into the conda environment, if it's not present (or if it's present without a version).

**Comment by tomasatdatabricks** [comment]
I think the MLflow is not necessarily required to be inside the user environment. Someone might use MLflow to package and run their project but do not use it within. And if it is used inside, it can be different version than executes the run. Injecting version of the "outside" MLFlow would break the reproducibility in some cases.  

But I can see how that is confusing. Maybe we should just . issue a warning if the versions do not match?

**Comment by aarondav** [comment]
I think at the very least, if the mlflow version in the conda environment is specified without a version, it should definitely be set to the one outside the environment. Anyway if you don't fix the version, it is not reproducible.

**Comment by tomasatdatabricks** [comment]
True. But no version gives you at least implicit >= current_version requirement. With fixing the version to the outside environment you could go several versions back.

**Comment by aarondav** [comment]
I don't think this is a complete argument; any conda env that includes mlflow via pip may eventually run into version incompatibility and caching problems. I think we should either explicitly not recommend this pattern (through docs, warnings, or errors), or give it different semantics.

One reason to enable this pattern by giving it different semantics is to give an easy way for users to keep up to date with mlflow, by tying it to the host version. On the other hand, it adds complexity (the behavior is not identical to loading the environment directly).

There may be alternative solutions to consider, like if we can break our conda cache if we identify that the pip dependencies are out of date, but that may take extra computational time which should be evaluated.

**Comment by andyk** [comment]
> but I would have to manually delete the mlflow-$sha conda environment in order to invalidate it.

On two separate occasions, I've had to do this to get things working again!

**Comment by tomasatdatabricks** [comment]
This is very similar to what we do for models. When serving models we do need mlflow to be installed inside of the model environment. We have decided not to install MLflow into the model environment by default. The model environment is expected to include mlflow. However, we did add a command line option to force-install mlflow into the environment. 

=> I think we should not change the current behavior but it may be a good idea to add a custom flag.
